### DISCUSSION TITLE
Piccola guida ai file DLG

### DISCUSSION THREAD

**Discussion by crazyLillo** [tesivo/tesivo_s_VTMB_italian_translation_for_Unofficial_Patch#57] [Generale]
In caso qualcuno approdasse qui, e volesse cimentarsi, ecco una piccola breve guida ai file DLG. (lo trovata nei file dell'SDK e ne ho tradotto un pezzettino.)

nota1 : NPC = Non player character (personaggio non giocante)
nota2 : PC = Player character (giocatore)

Se apri un file di dialogo, vedrai una tabella di 13 colonne definita con "{" e "}":

` { }{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }`
  
per semplificare la vista :

`{ 01}{ 02}{ 03}{04 }{05 }{06 }{07 }{08 }{09 }{10 }{11 }{12 }{13 }`

La PRIMA colonna è l'indice di riga. La prima riga dovrebbe sempre avere un valore di indice di 1:
 
 ```
     {1}{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }
      {2}{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }
      {3}{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }
      {14}{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }
```
    
Non tutti i numeri di riga devono essere contabilizzati (possono esserci dei salti per consentire di aggiungere risposte), tuttavia, tutte le righe dovrebbero essere visualizzate in ordine discendente.

La SECONDA colonna è ciò che dice l'NPC/PC se il protagonista è Maschio

```
      {1}{ Male : NPC Statement}{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }
      {2}{ Male : PC Response 1}{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }
      {3}{ Male : PC Response 2}{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }

```

La TERZA colonna è ciò che dice l'NPC/PC se il protagonista è una Femmina

```
       #  MALE   FEMALE
      {1}{Male}{ Female : NPC Statement}{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }
      {2}{Male}{ Female : PC Response 1}{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }
      {3}{Male}{ Female : PC Response 2}{ }{ }{ }{ }{ }{ }{ }{ }{ }{ }

```

La QUARTA colonna contiene un valore intero che reindirizza a un altra riga all'interno del file di dialogo se il PC sceglie quella risposta. 
Il carattere speciale " # " significa che è un'istruzione NPC

  ```
     #  MALE      FEMALE      NEXT
      {1}{NPC Male}{NPC Female}{ # }{ }{ }{ }{ }{ }{ }{ }{ }{ }
      {2}{PC Male }{PC Female }{ 5 }{ }{ }{ }{ }{ }{ }{ }{ }{ }
      {3}{PC Male }{PC Female }{ 7 }{ }{ }{ }{ }{ }{ }{ }{ }{ }
      {4}{        }{          }{   }{ }{ }{ }{ }{ }{ }{ }{ }{ }
      {5}{NPC Male}{NPC Female}{ # }{ }{ }{ }{ }{ }{ }{ }{ }{ }
      {6}{        }{          }{   }{ }{ }{ }{ }{ }{ }{ }{ }{ }
      {7}{NPC Male}{NPC Female}{ # }{ }{ }{ }{ }{ }{ }{ }{ }{ }
```

Durante la scansione delle risposte, il motore si arresta quando incontra una riga con un "#". La riga vuota non è necessaria, ma è stata aggiunta per migliorare la leggibilità.

La QUINTA colonna viene utilizzata per le condizioni. Puoi usare python o una delle condizioni speciali integrate.

Il motore supporta le operazioni logiche "and" utilizzando "and"

    `      {1}{M}{F}{5}{pc.armor==3 and npc.classname=="npc_VVampire"}`

Il motore di dialogo supporta le operazioni logiche "or" utilizzando "or", ma l'intera condizione deve essere racchiusa tra parentesi:

        `  {1}{M}{F}{5}{(pc.armor==3 or npc.classname=="npc_VVampire")}`

Non vedrai molte dichiarazioni "or" in tutte le finestre di dialogo perché puoi ottenere lo stesso effetto utilizzando 2 linee.

  ```
        {1}{M}{F}{5}{(pc.armor==3}
          {2}{M}{F}{5}{(npc.classname=="npc_VVampire"}

```

Ma cosa succede se entrambe le condizioni sono vere? Non vuoi che la stessa linea di dialogo appaia 2 volte. Per risolvere questo problema, potrebbe essere necessario aggiungere ulteriore logica.
Vedere anche gli esempi comuni seguenti per la funzione LimitSet().  LimitSet() è una funzione personalizzata che restituisce "true" solo una volta per un determinato set.

 ```
         {1}{M}{F}{5}{(pc.armor==3 and LimitSet(1,2)}
          {2}{M}{F}{5}{(npc.classname=="npc_VVampire" and LimitSet(2,2)}

```

La SESTA colonna ti consente di eseguire 1 linea di script Python. Se è un Istruzione NPC, lo script verrà eseguito quando l'NPC pronuncia l'istruzione. 
Se esso è all'interno di una scelta, lo script verrà eseguito solo se il giocatore sceglie quella scelta.

Esempio di Script: (In vamputils.py)

     ```
     def pcFollow(npc):
              npc.SetFollowerBoss("!player") 
 
          def pcNFollow(npc):
              npc.SetFollowerBoss("") 
```

Esempio di Dialogo:

   ```
       {1}{NPC}{NPC}{#}{                     }{              }{}{}{}{}{}{}{}
          {2}{sit}{sit}{1}{npc.IsFollowerOf(pc) }{pcNFollow(npc)}{}{}{}{}{}{}{}
          {3}{go }{go }{1}{!npc.IsFollowerOf(pc)}{pcFollow(npc) }{}{}{}{}{}{}{}
          {4}{end}{end}{}{                      }{              }{}{}{}{}{}{}{}
          {5}{   }{   }{ }{                     }{              }{}{}{}{}{}{}{}

```

Le COLONNE RIMANENTI (7-13) forniscono risposte specifiche basate sulla classe del PC.

      ```
Colonna 7 = Brujah
      Colonna 8 = Gangrel
      Colonna 9 = Nosferatu
      Colonna 10 = Toreador
      Colonna 11 = Tremere
      Colonna 12 = Ventrue
      Colonna 13 = Malkavian
```

Tieni presente che l'utilizzo di queste colonne non consente risposte specifiche per genere. Questo è uno dei motivi per cui non vengono più utilizzati durante il gioco. 
(Se vuoi usarli, puoi comunque SIMULARE  risposte in base al genere specifico, utilizzando condizionali di partenza. (lo vedremo in basso)

 AUTO-END

Il comando di fine automatico chiuderà automaticamente la finestra di dialogo se il comando è raggiunto. È incorporato nella sezione di risposta circondata da parentesi.

Sotto, non c'è altro che un solo AUTO-END nella colonna "Tutti", quindi la finestra di dialogo terminerà sempre dopo la dichiarazione dell'NPC.

         ```
    {201}{(STATEMENT MALE)}{STATEMENT FEMALE }{#}{ }{}{}{}{}{}{}{}{}
             {202}{(AUTO-END)}{}{0}{}{}{}{}{}{}{}{}{}
```
     

Nel secondo esempio di seguito, la conversazione si concluderà a meno che il PC è una femmina.

           ```
  {201}{(STATEMENT MALE)}{STATEMENT FEMALE }{#}{ }{}{}{}{}{}{}{}{}
             {202}{(AUTO-END)}{I Don't Know}{0}{}{}{}{}{}{}{}{}{}

```

 AUTO-LINK

Il comando di collegamento automatico si collegherà automaticamente a una nuova riga nella finestra di dialogo quando viene raggiunto.

Sotto, non c'è altro che un singolo AUTO-LINK nella colonna "Tutti", quindi la finestra di dialogo si collegherà alla successiva istruzione NPC.

   ```
          {201}{(STATEMENT MALE)}{STATEMENT FEMALE }{#}{ }{}{}{}{}{}{}{}{}
             {202}{(AUTO-LINK)}{}{211}{}{}{}{}{}{}{}{}{}

```

Ecco un esempio avanzato. Se l'npc sta seguendo il pc ed è femmina, vedrai una risposta "no". 
Se l'npc sta seguendo il pc ed è maschio, la finestra di dialogo salta a 211. 
Se l'npc NON sta seguendo il pc e il pc è maschio, ti autolinki al 221, altrimenti finisci.

     ```
        {201}{(STATEMENT MALE)}{STATEMENT FEMALE }{#}{ }{}{}{}{}{}{}{}{}
             {202}{(AUTO-LINK)}{No}{211}{pc.IsFollower()}{}{}{}{}{}{}{}{}
             {202}{(AUTO-LINK)}{}{221}{pc.IsMale()}{}{}{}{}{}{}{}{}
             {202}{(AUTO-END)}{}{0}{ }{}{}{}{}{}{}{}{}
```

CONDIZIONI DI PARTENZA

 
Nella parte inferiore della maggior parte dei file di dialogo, troverai molte righe contrassegnate come (CONDIZIONE DI PARTENZA). 
Quando viene caricata una finestra di dialogo, il motore di dialogo effettua la ricerca per un blocco di CONDIZIONI INIZIALI alla fine della finestra di dialogo. Sono valutate dall'alto verso il basso. Non appena la prima condizione restituisce " true", il motore viene reindirizzato alla prima linea degli NPC. 
Nota che non puoi concatenare le CONDIZIONI DI PARTENZA (non possono essere reindirizzate a un'altro blocco delle condizioni di partenza)

Esempio:

  
 ```
         {2000}{(STARTING CONDITION)}{}{10}{ not pc.IsMale()}{}{}{}{}{}{}{}{}
          {2001}{(STARTING CONDITION)}{}{20}{ pc.Charisma>2()}{}{}{}{}{}{}{}{}

```

Sopra, se il pc è femmina, vai alla linea di dialogo a 10. Se il pc è maschio e ha carisma decente, vai alla linea 20. 
Se nessuna delle condizioni di partenza  è "true", il dialogo tenta di avviarsi all'inizio del  file (riga 1). 
Usando un sistema come questo, potresti progettare delle risposte specifiche di un clan per entrambi i sessi.

Il codice inserito nella sesta colonna NON viene eseguito con condizioni di partenza TUTTAVIA, le condizioni supportano Python e non c'è nulla che ti impedisca di eseguire codice nel blocco condizionale.

Questo è utile principalmente per collocare gli npcs prima dell'avvio delle finestre di dialogo:

```
         {200}{(STARTING CONDITION)}{}{0 }{ mySetup(npc)   }{}{}{}{}{}{}{}{}
         {201}{(STARTING CONDITION)}{}{10}{ not pc.IsMale()}{}{}{}{}{}{}{}{}
         {202}{(STARTING CONDITION)}{}{20}{ pc.Charisma>2()}{}{}{}{}{}{}{}{}

```
```
      (In Vamputils.py)
      def mySetup(npc):
          # do whatever
          return 0

```

Si noti che il numero di riga della prima CONDIZIONE INIZIALE è 0. Quando questo è il caso, la condizione iniziale viene valutata ma il suo risultato è ignorato e va avanti.
In altre parole, non importa se  mySetup(npc) restituisce 0 o 1, tuttavia restituiamo 0 per essere al sicuro.

**Comment by tesivo** [comment]
Grande aggiunta per chiunque si volesse cimentare nella traduzione, o perchè no, nello sviluppo della unofficial patch stessa :)
